### PR TITLE
[libc++][span] P2447R4: `std::span` over an initializer list

### PR DESCRIPTION
**Pull request by H-G-Hristov**
Implements: https://wg21.link/P2447R6
- https://eel.is/c++draft/span.syn
- https://eel.is/c++draft/span.overview
- https://eel.is/c++draft/span.cons
- https://eel.is/c++draft/diff

### CODE REVIEW

File: libcxx/test/std/containers/views/views.span/span.cons/iterator_len.verify.cpp [hunk 1]
```diff
@@ -13,22 +13,36 @@
...
+void test() {
   // explicit constructor necessary
   int arr[] = {1, 2, 3};
   createImplicitSpan<int, 1>(arr, 3);
 
-  std::span<int> sp = {0, 0}; // expected-error {{no matching constructor for initialization of 'std::span<int>'}}
-  std::span<int, 2> sp2 = {0, 0}; // expected-error {{no matching constructor for initialization of 'std::span<int, 2>'}}
-  std::span<const int> csp = {0, 0}; // expected-error {{no matching constructor for initialization of 'std::span<const int>'}}
-  std::span<const int, 2> csp2 = {0, 0}; // expected-error {{no matching constructor for initialization of 'std::span<const int, 2>'}}
+  // expected-error@+1 {{no matching constructor for initialization of 'std::span<int>'}}
```
Comment (mordante) [inline libcxx/test/std/containers/views/views.span/span.cons/iterator_len.verify.cpp:33]: Can you add tests for verify all changes listed in Annex C?

File: libcxx/include/span [hunk 2]
```diff
@@ -228,6 +229,14 @@ public:
     requires(_Sz == 0)
   _LIBCPP_HIDE_FROM_ABI constexpr span() noexcept : __data_{nullptr} {}
 
+#  if _LIBCPP_STD_VER >= 26
+  _LIBCPP_HIDE_FROM_ABI constexpr explicit span(std::initializer_list<value_type> __il)
+    requires is_const_v<element_type>
+      : __data_{__il.begin()} {
+    _LIBCPP_ASSERT_INTERNAL(_Extent == __il.size(), "Size mismatch in span's constructor _Extent != __il.size().");
```
Comment (mordante) [inline libcxx/include/span:236]:
```suggestion
    _LIBCPP_ASSERT_VALID_INPUT_RANGE(_Extent == __il.size(), "Size mismatch in span's constructor _Extent != __il.size().");
```

File: libcxx/test/std/containers/views/views.span/span.cons/iterator_len.verify.cpp [hunk 3]
```diff
@@ -13,22 +13,36 @@
...
+  std::span<int> sp = {0, 0};
+  // expected-error@+1 {{no matching constructor for initialization of 'std::span<int, 2>'}}
+  std::span<int, 2> sp2 = {0, 0};
+#if TEST_STD_VER >= 26
+  // No error in C++26
+  std::span<const int> csp = {0, 0};
+  // expected-error@+1 {{chosen constructor is explicit in copy-initialization}}
```
Comment (mordante) [inline libcxx/test/std/containers/views/views.span/span.cons/iterator_len.verify.cpp:40]: We typically test explicitness without a verify test.
The main function in `test/std/input.output/file.streams/fstreams/ifstream.cons/path.pass.cpp`.

In general we prefer to test things without verify tests. These are only tested by us not by other standard libraries using our test suite. They are sensitive to changes in Clang diagnostics.

**Review by mordante — CHANGES_REQUESTED** [review]
Mostly looks good, some minor issues.

File: libcxx/test/std/containers/views/views.span/span.cons/iterator_len.verify.cpp (hunk 3 above)
Comment (H-G-Hristov) [reply libcxx/test/std/containers/views/views.span/span.cons/iterator_len.verify.cpp]: Thanks! I'll update the tests.

**Comment by H-G-Hristov** [comment]
Thanks. The documentation build fails due to some CMake error consistently otherwise I believe I addressed all comments.

File: libcxx/test/std/containers/views/views.span/span.cons/initializer_list.assert.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  TEST_LIBCPP_ASSERT_FAILURE(
+      (std::span<const int, 4>({1, 2, 3, 9084, 5})), "Size mismatch in span's constructor _Extent != __il.size().");
+  TEST_LIBCPP_ASSERT_FAILURE((std::span<const int, 4>(std::initializer_list<int>{1, 2, 3, 9084, 5})),
+                             "Size mismatch in span's constructor _Extent != __il.size().");
+
+  return true;
+}
+
+int main(int, char**) {
+  assert(test());
```
Comment (mordante) [inline libcxx/test/std/containers/views/views.span/span.cons/initializer_list.assert.pass.cpp:35]: We typically don't do this for run-time tests. We do it for compile-time tests so we can use `static_assert`s. In this case I would remove the `test` function and move the test code in `main`.

I overlooked this in the previous review.

File: libcxx/test/std/containers/views/views.span/span.cons/initializer_list.pass.cpp [hunk 5]
```diff
@@ -5,39 +5,134 @@
...
+constexpr int four(std::span<const std::any> sp) { return sp.size(); }
+
+bool test_P2447R4_annex_c_examples() {
+  // 1. Overload resolution is affected
+  // --> tested in "initializer_list.verify.cpp"
+
+  // 2. The `initializer_list` ctor has high precedence
+  // --> tested in "initializer_list.verify.cpp"
+
+  // 3. Implicit two-argument construction with a highly convertible value_type
```
Comment (mordante) [inline libcxx/test/std/containers/views/views.span/span.cons/initializer_list.pass.cpp:107]: nice! thanks for referring to the other test too!

**Review by mordante — APPROVED** [review]
> Thanks. The documentation build fails due to some CMake error consistently otherwise I believe I addressed all comments.
I think I fixed this a few days ago.

LGTM modulo 1 nit.

**Comment by H-G-Hristov** [comment]
> > Thanks. The documentation build fails due to some CMake error consistently otherwise I believe I addressed all comments.
> > I think I fixed this a few days ago.
> 
> LGTM modulo 1 nit.

Thanx!

**Pull request merged**